bindings/matrix-sdk-ffi/src/room_preview.rs

```diff
@@ -1,9 +1,87 @@
-use matrix_sdk::{room_preview::RoomPreview as SdkRoomPreview, RoomState};
+use matrix_sdk::{
+    room_preview::{
+        JoinRoomPreviewAction as SdkJoinRoomPreviewAction, LeaveRoomPreviewAction as SdkLeaveRoomPreviewAction, RoomPreview as SdkRoomPreview,
+        RoomPreviewActions,
+    },
+    Client, RoomState,
+};
 use ruma::space::SpaceRoomJoinRule;
 
-/// The preview of a room, be it invited/joined/left, or not.
+use crate::{client::JoinRule, error::ClientError, room::Membership};
+
 #[derive(uniffi::Record)]
 pub struct RoomPreview {
+    pub info: RoomPreviewInfo,
+    pub room_preview_actions: RoomPreviewActions,
+    pub(crate) sdk_info: matrix_sdk::room_preview::RoomPreview,
+}
+
+#[derive(uniffi::Enum)]
+pub enum RoomPreviewAction {
+    Invited { join: JoinRoomPreviewAction, leave: LeaveRoomPreviewAction },
+    Knocked { leave: LeaveRoomPreviewAction },
+}
+
+#[derive(uniffi::Object)]
+pub struct JoinRoomPreviewAction {
+    inner: SdkJoinRoomPreviewAction,
+}
+
+#[matrix_sdk_ffi_macros::export]
+impl JoinRoomPreviewAction {
+    async fn run(&self) {
+        self.inner.run().unwrap()
```

Comment:
Let's not `unwrap()` at the FFI layer, this will cause unnecessary panics that we need to handle at the last minute.

Reply:
Removed in https://github.com/matrix-org/matrix-rust-sdk/pull/4152/commits/0fcfd84311e2047825763aeae41cc210d3007041.